pkgs/build-support/setup-hooks/make-binary-wrapper.sh

```diff
@@ -0,0 +1,198 @@
+            uses_prefix=1
+            uses_concat3=1
+            n=$((n + 3))
+        elif [[ "$p" == "--suffix" ]]; then
+            cmd=$(setEnvSuffix "${params[$((n + 1))]}" "${params[$((n + 2))]}" "${params[$((n + 3))]}")
+            main="$main    $cmd"$'\n'
+            uses_suffix=1
+            uses_concat3=1
+            n=$((n + 3))
+        elif [[ "$p" == "--add-flags" ]]; then
+            flags="${params[$((n + 1))]}"
+            flagsBefore="$flagsBefore $flags"
+            n=$((n + 1))
+        elif [[ "$p" == "--argv0" ]]; then
+            argv0=$(escapeStringLiteral "${params[$((n + 1))]}")
+            n=$((n + 1))
+        else
+            # Using an error macro, we will make sure the compiler gives an understandable error message
+            printf "%s\n" "    #error makeCWrapper did not understand argument ${p}"
+        fi
+    done
+    [ -z "$flagsBefore" ] || main="$main"${main:+$'\n'}$(addFlags $flagsBefore)$'\n'$'\n'
+    main="$main    argv[0] = \"${argv0:-${executable}}\";"$'\n'
+    main="$main    return execv(\"${executable}\", argv);"$'\n'
+
+    printf "%s\n" "#include <unistd.h>"
+    printf "%s\n" "#include <stdlib.h>"
+    [ -z "$uses_concat3" ] || printf "%s\n" "#include <string.h>"
+    [ -z "$uses_concat3" ] || printf "\n%s\n" "$(concat3Fn)"
+    [ -z "$uses_prefix" ]  || printf "\n%s\n" "$(setEnvPrefixFn)"
+    [ -z "$uses_suffix" ]  || printf "\n%s\n" "$(setEnvSuffixFn)"
+    printf "\n%s" "int main(int argc, char **argv) {"
+    printf "\n%s" "$main"
+    printf "%s\n" "}"
+}
+
+addFlags() {
+    local result n flag flags
+    local var="argv_tmp"
+    flags=("$@")
+    for ((n = 0; n < ${#flags[*]}; n += 1)); do
+        flag=$(escapeStringLiteral "${flags[$n]}")
+        result="$result    $var[$((n+1))] = \"$flag\";"$'\n'
+    done
+    printf "    %s\n" "char **$var = malloc(sizeof(*$var) * ($((n+1)) + argc));"
+    printf "    %s\n" "$var[0] = argv[0];"
+    printf "%s" "$result"
+    printf "    %s\n" "for (int i = 1; i < argc; ++i) {"
+    printf "    %s\n" "    $var[$n + i] = argv[i];"
+    printf "    %s\n" "}"
+    printf "    %s\n" "$var[$n + argc] = NULL;"
+    printf "    %s\n" "argv = $var;"
+}
+
+# prefix ENV SEP VAL
+setEnvPrefix() {
+    local env=$(escapeStringLiteral "$1")
+    local sep=$(escapeStringLiteral "$2")
+    local val=$(escapeStringLiteral "$3")
+    printf "%s" "set_env_prefix(\"$env\", \"$sep\", \"$val\");"
+}
+
+# suffix ENV SEP VAL
+setEnvSuffix() {
+    local env=$(escapeStringLiteral "$1")
+    local sep=$(escapeStringLiteral "$2")
+    local val=$(escapeStringLiteral "$3")
+    printf "%s" "set_env_suffix(\"$env\", \"$sep\", \"$val\");"
+}
+
+# setEnv KEY VALUE
+setEnv() {
+    local key=$(escapeStringLiteral "$1")
+    local value=$(escapeStringLiteral "$2")
+    printf "%s" "putenv(\"$key=$value\");"
+}
+
+# setDefaultEnv KEY VALUE
+setDefaultEnv() {
+    local key=$(escapeStringLiteral "$1")
+    local value=$(escapeStringLiteral "$2")
```

Comment:
shellcheck will be pretty unhappy about all of those. Please check the script with it.

Reply:
There's no issue here. Look for "assignment statements" in https://www.gnu.org/software/bash/manual/bash.html#Shell-Parameters

Shellcheck does warn about masking return values, but it's not like `escapeStringLiteral` can fail.

Reply:
> There's no quoting issue here.

I never talked about quoting. I was referring to return value masking. Use shellcheck to get the exact details.

Reply:
Yeah, I get this warning from shellcheck:
```
Declare and assign separately to avoid masking return values.shellcheck(SC2155)
```

If I understand correctly, I need to change from

```bash
local key=$(escapeStringLiteral "$1")
local value=$(escapeStringLiteral "$2")
```

to:

```bash
local key value
key=$(escapeStringLiteral "$1")
value=$(escapeStringLiteral "$2")
```